src/track/beatutils.cpp

```diff
@@ -29,20 +31,177 @@ const int kHistogramDecimalPlaces = 2;
 const double kHistogramDecimalScale = pow(10.0, kHistogramDecimalPlaces);
 const double kBpmFilterTolerance = 1.0;
 
+QMap<int, double> BeatUtils::findTempoChanges(
+    QMap<double, int> tempoFrequency, QList<double> tempoList) {
+
+    auto sortedTempoList = tempoList;
+    std::sort(sortedTempoList.begin(), sortedTempoList.end());
+    // We have to make sure we have odd numbers
+    if (!(sortedTempoList.size() % 2) and sortedTempoList.size() > 1) {
+        sortedTempoList.pop_back();
+    }
+    // Since we use the median as a guess first and last tempo
+    // And it can not have values outside from tempoFrequency
+    const double median = computeSampleMedian(sortedTempoList);
+    // The analyzer sometimes detect false beats that generate outliers
+    // values for the tempo so we use a median filter to remove them
+    MovingMedian filterTempo(5); // 5 is the lenght our window
+    int currentBeat = 0, lastBeatChange = 0;
+    QMap<int, double> stableTemposAndPositions;
+    stableTemposAndPositions[lastBeatChange] = median;
+    // Here we are going to track the tempo changes over the track
+    for (double tempo : tempoList) {
+        double newStableTempo = filterTempo(tempo);
+        // The analyzer has some jitter that causes a steady beat to fluctuate around
+        // the correct value so we don't consider tempos +-1 from that as changes
+        if (newStableTempo == stableTemposAndPositions.last()) {
+            ; // no-op - action only in else
```

Comment:
If you initialize currentBeat to -1 and increment it before the if case, you can use continue here. Not sure if that would be better.

Reply:
That would be cleaner I guess